test/unit/app/controllers/permissions/permissions-controller-test.js

```diff
@@ -0,0 +1,1392 @@
+      )
+
+      // approve out of order
+      await permController.approvePermissionsRequest(request2, ACCOUNT_ARRAYS.b)
+      // add a non-existing request to the mix
+      await permController.approvePermissionsRequest(request3, ACCOUNT_ARRAYS.c)
+      await permController.approvePermissionsRequest(request1, ACCOUNT_ARRAYS.a)
+
+      assert.deepEqual(
+        perms1, PERMS.finalizedRequests.eth_accounts(ACCOUNT_ARRAYS.a),
+        'first request produced expected approved permissions'
+      )
+
+      assert.deepEqual(
+        perms2, PERMS.finalizedRequests.eth_accounts(ACCOUNT_ARRAYS.b),
+        'second request produced expected approved permissions'
+      )
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty after approvals',
+      )
+    })
+  })
+
+  describe('rejectPermissionsRequest', function () {
+
+    let permController, mockRequestUserApproval
+
+    beforeEach(async function () {
+      permController = initPermController()
+      mockRequestUserApproval = getMockRequestUserApprovalFunction(
+        permController
+      )
+    })
+
+    it('does nothing if called on non-existing request', async function () {
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty on init',
+      )
+
+      assert.doesNotReject(
+        permController.rejectPermissionsRequest(REQUEST_IDS.a),
+        'should not throw on non-existing request'
+      )
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals still empty after request',
+      )
+    })
+
+    it('rejects single existing request', async function () {
+
+      assert.rejects(
+        mockRequestUserApproval(REQUEST_IDS.a),
+        ERRORS.rejectPermissionsRequest.rejection(),
+        'rejects as expected'
+      )
+
+      await permController.rejectPermissionsRequest(REQUEST_IDS.a)
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty after rejection',
+      )
+    })
+
+    it('rejects requests regardless of order', async function () {
+
+      assert.rejects(
+        mockRequestUserApproval(REQUEST_IDS.b),
+        ERRORS.rejectPermissionsRequest.rejection(),
+        'rejects as expected'
+      )
+
+      assert.rejects(
+        mockRequestUserApproval(REQUEST_IDS.c),
+        ERRORS.rejectPermissionsRequest.rejection(),
+        'rejects as expected'
+      )
+
+      // reject out of order
+      await permController.rejectPermissionsRequest(REQUEST_IDS.c)
+      // add a non-existing request to the mix
+      await permController.rejectPermissionsRequest(REQUEST_IDS.a)
+      await permController.rejectPermissionsRequest(REQUEST_IDS.b)
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty after approval',
+      )
+    })
+  })
+
+  describe('_requestPermissions', function () {
```

Comment:
I disagree, and here's the argument: https://stackoverflow.com/a/105209

Reply:
That answer reads as "I write them because they are helpful" to me, e.g. it doesn't say a great deal. Whereas the comments and other answers provide a number of different counter-arguments.

Ultimately unit tests are supposed to help us write software, right? We can test our assumptions and ensure those assumptions hold true even as we make unrelated changes or refactor. If we test the internal implementation details, we're making future changes more difficult by cementing everything in place. It becomes a make-work project that consumes a great deal of time and has dubious benefit. The greatest strength of unit tests is in allowing fearless refactors, and this is the opposite of that 🤷‍♂.

This is kinda what I meant about refactoring the code-under-test being a necessity to writing truly useful unit tests, because stumbling blocks like this are generally a sign that things could be broken out and made simpler. e.g. if an internal method is complex enough that testing it _indirectly_ is painful, then maybe it should be a separate module, or maybe some of the complexity isn't required and should be removed.

I've been the person on the team to argue for 100% test coverage and for testing internal methods, but in the end it was painful to maintain and yielded minimal benefits.

There are of course exceptions. e.g. maybe a function actually _is_ part of _a_ public interface, just not the _module_ interface. Or maybe the function should probably be split out into a separate module, but that work has been delayed due to the complexity. But as a general rule I don't think testing internals has any benefit.

Reply:
Well, what really does my position in is that only one line became uncovered after I commented out the tests for `_requestPermissions` and adding/removing pending approvals. I moved one case to the miscellaneous tests, and that's that. Done in 9c03bf2